src/shared/components/editableSpan/EditableSpan.tsx

```diff
@@ -24,45 +36,109 @@ export default class EditableSpan extends React.Component<IEditableSpanProps, {}
-            evt.preventDefault();
-            return;
-        }
 
         if (this.props.numericOnly) {
             const regex = /^\d$/;
             if(!regex.test(newKey)) {
                 evt.preventDefault();
             }
+        } else if (this.props.isPercentage) {
+            const regex = /^[%\d]$/;
+            if(!regex.test(newKey)) {
+                evt.preventDefault();
+            }
+        }
+
+        // By this point, the new character is acceptable to be typed, if not past max length already
+
+        // If at max length...
+        if (this.props.maxChars &&
+            (this.spanElt.innerText.length === this.props.maxChars)
+        ) {
+            // ...dont type unless at least one character selected to be replaced
+            const selection = window.getSelection();
+            if (!selection.containsNode(this.spanElt, true) || !selection.toString().length) {
+                evt.preventDefault();
+            }
+        }
+    }
+
+    private onKeyUp(evt:React.KeyboardEvent<HTMLSpanElement>) {
+        this.props.onKeyUp && this.props.onKeyUp(evt);
+
+        if (this.props.maxChars &&
+            (this.spanElt.innerText.length > this.props.maxChars)
+        ) {
+            // If something has gone wrong and the input length is longer than allowed, truncate it
+            this.spanElt.innerText = this.spanElt.innerText.substring(this.spanElt.innerText.length-this.props.maxChars);
         }
     }
 
     private onBlur(evt:React.FocusEvent<HTMLSpanElement>) {
         this.props.onBlur && this.props.onBlur(evt);
 
-        if (evt.currentTarget.innerText.length > 0) {
-            this.props.setValue(evt.currentTarget.innerText);
+        const inputText = evt.currentTarget.innerText;
+        this.setText(this.props.value);
+
+        if (inputText.length > 0 && (inputText !== this.props.value || this.enterPressedSinceLastBlur)) {
+            this.dirty = true;
+            this.props.setValue(inputText);
+        }
+        this.enterPressedSinceLastBlur = false;
+    }
+
+    private setText(text:string) {
+        this.spanElt.innerText = text;
+    }
+
+    componentDidMount() {
+        this.setText(this.props.value);
+    }
+
+    componentWillReceiveProps(nextProps:IEditableSpanProps) {
+        if (this.dirty || (nextProps.value !== this.props.value)) {
+            this.dirty = false;
+            this.setText(nextProps.value);
         }
-        evt.currentTarget.innerText = this.props.value;
     }
 
     render() {
-        const {contentEditable, onKeyPress, onBlur,
+        const {contentEditable, onKeyPress, onKeyUp, onBlur,
             value, setValue, maxChars,
-            numericOnly, ...spanProps} = this.props;
+            numericOnly, isPercentage,
+            textFieldAppearance,
+            style, ...spanProps} = this.props;
+        let spanStyle = {};
+        if (this.props.textFieldAppearance) {
```

Comment:
rather than complicating the component with this kind of conditional style stuff, why not just create and apply a css class?

Reply:
the actual reason why is that I'm not as comfortable/familiar with how scss module magic works. But I can do that.

Reply:
You wouldn't actually have to use css modules.  I would recommend that EditableSpan accept a custom class name.  Then for a particular instance, you define that class in the feature's scss file.  Let me know if you want more detail.